File: multi_machine/c_multi_machine_generator/gerador.c
/* Gerador dos exemplos do artigo do European */
#include<stdio.h>
#include<string.h>
#include<math.h>
// #include<sys/times.h>
#include<sys/types.h>


#define  MAX_T		   20
#define  MAX_M		    8
#define  MAX_N		   52

double    maq;
double    hpdc[MAX_N], dpdc[MAX_T][MAX_N];
double    fpdc[MAX_M][MAX_N], spdc[MAX_M][MAX_N];
double    cpdc[MAX_M][MAX_N];
double    bpdc[MAX_M][MAX_N];
double	  I[MAX_T][MAX_N],
          x[MAX_T][MAX_M][MAX_N];




double unif(long int *seed, double low, double high)
{
  double   unif_ret;
  long int m,a,b,c, k;
  double   value_0_1;

  m = 2147483647;
  a = 16807;
  b = 127773;
  c = 2836;

  k = *seed/b;
  *seed = a * (*seed % b) - k * c;
  if (*seed <0)
    *seed = *seed + m;
  value_0_1 = (double) *seed/m;
  unif_ret = low + value_0_1 * (high - low);
  return (unif_ret);
}



int Gera_dados (char nome[10], int NN, int MM, int TT)
{
  int    narq          ;
  char   fname_dat[30] ;
  FILE  *dat           ;
  FILE  *arq           ;
  char   fname_ent[15] ;
  double L_min_s, L_max_s,
         L_min_c, L_max_c,
         L_min_h, L_max_h,
         L_min_d, L_max_d,
         L_min_f, L_max_f,
         L_min_b, L_max_b,
         fracao, auxf;
  long int    seed;
  int    aux1, ii, jj, tt;
  double max_f, max_b;

/* Le do arquivo gdado.dat os parametros que serao utilizados na geracao
   dos problemas testes                                                   */

  sprintf(fname_ent,"../gdado.dat");
  if ((arq = fopen (fname_ent, "r")) == NULL){
    printf ("Erro na abertura do arquivo de dados !!!\n");
    return(1);
  }

  fscanf( arq, "%*s" );
  fscanf( arq, "%*s %*s %lf", &L_min_s );
  fscanf( arq, "%*s %*s %lf", &L_max_s );
  fscanf( arq, "%*s %*s %lf", &L_min_c );
  fscanf( arq, "%*s %*s %lf", &L_max_c );
  fscanf( arq, "%*s %*s %lf", &L_min_h );
  fscanf( arq, "%*s %*s %lf", &L_max_h );
  fscanf( arq, "%*s %*s %lf", &L_min_d );
  fscanf( arq, "%*s %*s %lf", &L_max_d );

  fscanf (arq, "%*s");
  fscanf( arq, "%*s %*s %lf", &L_min_b );
  fscanf( arq, "%*s %*s %lf", &L_max_b );
  fscanf( arq, "%*s %*s %lf", &L_min_f );
  fscanf( arq, "%*s %*s %lf", &L_max_f );

//  fscanf( arq, "%*s %*s %d" , &L_NP);

  fclose(arq);

  /* inicio da geracao dos dados - gera 10 problemas para cada NxMxT dado */

  if ((arq = fopen ("../semente.dat", "r")) == NULL){
    printf ("Erro na abertura do arquivo de dados !!!\n");
    return(1);
  }

  for(narq=0;narq<10;narq++){

    fscanf(arq, "%*s %*s %d",&seed);
    sprintf(fname_dat,"%s%d_%d_%d_%d.dat",nome,narq,NN,MM,TT);

    if ((dat = fopen (fname_dat, "w")) == NULL) {
     printf ("Erro na abertura do arquivo de dados !!!\n");
     return(1);
    }

    /* geracao dos dados */

    for(ii=1;ii<=NN;ii++){
      hpdc[ii] = unif(&seed,L_min_h,L_max_h);
      for(jj=1;jj<=MM;jj++){
        cpdc[jj][ii] = unif(&seed,L_min_c,L_max_c);
        spdc[jj][ii] = unif(&seed,L_min_s,L_max_s);
      }
    }

    for(ii=1;ii<=NN;ii++)
      for(jj=1;jj<=MM;jj++){
        bpdc[jj][ii] = unif(&seed,L_min_b,L_max_b);
        fpdc[jj][ii] = unif(&seed,L_min_f,L_max_f);
      }


    for(ii=1;ii<=NN;ii++)
      for(tt=1;tt<=TT;tt++)
        dpdc[tt][ii] = 1.0* ((int) (unif(&seed,L_min_d,L_max_d)));

  /* Calculo da capacidade */

  auxf = 0.0;

  for(ii=1;ii<=NN;ii++)
    for(tt=1;tt<=TT;tt++)
      if (dpdc[tt][ii]>0.0){
        fracao = dpdc[tt][ii]/MM;
        for(jj=1;jj<=MM;jj++)
          auxf += fpdc[jj][ii] + bpdc[jj][ii]*fracao;
      }

  aux1 = (int)(auxf/(MM*TT)); /*politica lote por lote */

  if ((auxf - aux1) > 0.5)
    maq = 1.0*aux1+1.0;
  else
    maq = 1.0*aux1;

  /* Fim do calculo da capacidade */


  /* Escrevendo no arquivo de saida */

  fprintf(dat,"%d %d\n",NN, TT);
  fprintf(dat,"%d\n",MM);

  fprintf(dat,"%10.0lf\n",maq);

  for(jj=1;jj<=MM;jj++)
    for(ii=1;ii<=NN;ii++){
      fprintf(dat,"%5.1lf %5.1lf ",bpdc[jj][ii],fpdc[jj][ii]);
      fprintf(dat,"%5.1lf %5.1lf \n",spdc[jj][ii],cpdc[jj][ii]);
    }
  for(ii=1;ii<=NN;ii++)
    fprintf(dat,"%5.1lf ",hpdc[ii]);
  fprintf(dat,"\n");


  if(NN<=15){
   for(tt=1;tt<=TT;tt++){
    for(ii=1;ii<=NN;ii++)
     fprintf(dat,"%5d ",(int)(dpdc[tt][ii]));
    fprintf(dat,"\n");
   }
  }
  else{
   for(tt=1;tt<=TT;tt++){
    for(ii=1;ii<=15;ii++)
     fprintf(dat,"%5d ",(int)(dpdc[tt][ii]));
    fprintf(dat,"\n");
   }

   for(tt=1;tt<=TT;tt++){
    for(ii=16;ii<=NN;ii++)
     fprintf(dat,"%5d ",(int)(dpdc[tt][ii]));
    fprintf(dat,"\n");
   }
  }

  fclose(dat);
  }
  fclose(arq);
}



void roda_problemas(char tipo[10],int inicio, int fim, char nome[10], double tipocap, double stcost, double sttime)
{
  int    vetorN[5];
  int    erro, TT, MM, id, NN;

  vetorN[1] = 6; vetorN[2] = 12; vetorN[3] = 25; vetorN[4] = 50;

  for(TT=6;TT<=18;TT=TT+6)
    for(MM=2;MM<=6;MM=MM+2)
      for(id=1;id<=4;id++){
        NN = vetorN[id];
        erro = Gera_dados(nome,NN,MM,TT);
      }
}


int main(void)
{
 FILE *saida;

 roda_problemas("CASBTB",0,9,"ABB0",0.9, 1.0,1.0);

 roda_problemas("CASATB",0,9,"AAB0",0.9, 10.0,1.0);
 
 roda_problemas("CASBTA",0,9,"ABA0",0.9, 1.0,1.5);

 roda_problemas("CASATA",0,9,"AAA0",0.9, 10.0,1.5);
 
 roda_problemas("CNSBTB",0,9,"NBB0",1.0, 1.0,1.0);

 roda_problemas("CNSATB",0,9,"NAB0",1.0, 10.0,1.0);

 roda_problemas("CNSBTA",0,9,"NBA0",1.0, 1.0,1.5);

 roda_problemas("CNSATA",0,9,"NAA0",1.0, 10.0,1.5);

 return(0);
}

File: multi_machine/python_multi_machine_generator/gerador.py
def unif(seed, low, high):
    """Gerador que utiliza método da congruência linear"""
    m = 2147483647
    a = 16807
    b = 127773
    c = 2836

    k = seed // b
    seed = a * (seed % b) - k * c
    if seed < 0:
        seed += m
    value_0_1 = seed / m
    unif_ret = low + value_0_1 * (high - low)
    return unif_ret, seed


def Gera_dados(nome, NN, MM, TT):
    """Função principal de leitura dos parâmetros e contrução dos arquivos .dat"""
    # Read parameters from "gdado.dat"
    fname_ent = "../gdado.dat"
    try:
        with open(fname_ent, "r") as arq:
            lines = arq.readlines()
    except IOError:
        print("Erro na abertura do arquivo de dados !!!")
        return 1

    # Initialize parameters
    params = {}
    # Process each line to extract parameters
    for line in lines:
        line = line.strip()
        if not line or '=' not in line:
            continue  # Skip empty lines and lines without '='
        tokens = line.split('=')
        key = tokens[0].strip()
        value = float(tokens[1].strip())
        params[key] = value

    # Assign parameters
    L_min_s = params.get('mincustosetup', 0)
    L_max_s = params.get('maxcustosetup', 0)
    L_min_c = params.get('mincustoprod', 0)
    L_max_c = params.get('maxcustoprod', 0)
    L_min_h = params.get('mincustoestoq', 0)
    L_max_h = params.get('maxcustoestoq', 0)
    L_min_d = params.get('mindemanda', 0)
    L_max_d = params.get('maxdemanda', 0)
    L_min_b = params.get('mintempoprod', 0)
    L_max_b = params.get('maxtempoprod', 0)
    L_min_f = params.get('mintemposetup', 0)
    L_max_f = params.get('maxtemposetup', 0)
    # If any parameter is missing, print an error
    required_params = ['mintemposetup', 'maxtemposetup', 'mincustoprod', 'maxcustoprod',
                       'mincustoestoq', 'maxcustoestoq', 'mindemanda', 'maxdemanda',
                       'mintempoprod', 'maxtempoprod', 'mincustosetup', 'maxcustosetup']
    for param in required_params:
        if param not in params:
            print(f"Missing parameter '{param}' in gdado.dat")
            return 1

    # Open "semente.dat" and read seeds
    try:
        with open("../semente.dat", "r") as arq:
            seed_lines = arq.readlines()
    except IOError:
        print("Erro na abertura do arquivo de dados !!!")
        return 1

    for narq in range(10):
        # Read seed
        tokens = seed_lines[narq].split('=')
        seed = int(tokens[-1].strip())

        fname_dat = f"{nome}{narq}_{NN}_{MM}_{TT}.dat"

        try:
            dat = open(fname_dat, "w")
        except IOError:
            print("Erro na abertura do arquivo de dados !!!")
            return 1

        # Generate data
        NN_index = NN  # To match C's 1-based indexing
        MM_index = MM
        TT_index = TT

        hpdc = [0.0] * NN
        cpdc = [ [0.0]*NN for _ in range(MM)]
        spdc = [ [0.0]*NN for _ in range(MM)]
        bpdc = [ [0.0]*NN for _ in range(MM)]
        fpdc = [ [0.0]*NN for _ in range(MM)]
        dpdc = [ [0.0]*NN for _ in range(TT)]

        for ii in range(NN):
            hpdc[ii], seed = unif(seed, L_min_h, L_max_h)
            for jj in range(MM):
                cpdc[jj][ii], seed = unif(seed, L_min_c, L_max_c)
                spdc[jj][ii], seed = unif(seed, L_min_s, L_max_s)

        for ii in range(NN):
            for jj in range(MM):
                bpdc[jj][ii], seed = unif(seed, L_min_b, L_max_b)
                fpdc[jj][ii], seed = unif(seed, L_min_f, L_max_f)

        for ii in range(NN):
            for tt in range(TT):
                value, seed = unif(seed, L_min_d, L_max_d)
                dpdc[tt][ii] = 1.0 * int(value)

        # Calculate capacity
        auxf = 0.0

        for ii in range(NN):
            for tt in range(TT):
                if dpdc[tt][ii] > 0.0:
                    fracao = dpdc[tt][ii] / MM
                    for jj in range(MM):
                        auxf += fpdc[jj][ii] + bpdc[jj][ii] * fracao

        aux1 = int(auxf / (MM * TT))

        if (auxf - aux1) > 0.5:
            maq = aux1 + 1.0
        else:
            maq = aux1 * 1.0

        # FORMATAÇÃO DO ARQUIVO DE OUTPUT
        # Número de produtos | Número de períodos
        print(f"{NN} {TT}", file=dat)
        # Número de máquinas
        print(f"{MM}", file=dat)
        # Capacidade calculada acima
        print(f"{maq:10.0f}", file=dat)

        # Tempo de produção | Tempo setup | Custo de setup | Custo de produção
        for jj in range(MM):
            for ii in range(NN):
                print(f"{bpdc[jj][ii]:5.1f} {fpdc[jj][ii]:5.1f} {spdc[jj][ii]:5.1f} {cpdc[jj][ii]:5.1f}", file=dat)

        # Linha com custos de armazenagem
        for ii in range(NN):
            print(f"{hpdc[ii]:5.1f} ", end='', file=dat)
        print("", file=dat)
        
        # Demanda com 1 período por linha
        for tt in range(TT):
            for ii in range(NN):
                print(f"{int(dpdc[tt][ii]):5d} ", end='', file=dat)
            print("", file=dat)

        dat.close()


def roda_problemas(tipo, inicio, fim, nome, tipocap, stcost, sttime):
    """Define o número de produtos (NN), períodos (TT) e máquinas (MM) e chama a função Gera_dados"""
    vetorN = [0, 6, 12, 25, 50]
    for TT in range(6, 19, 6):
        for MM in range(2, 7, 2):
            for id in range(1, 5):
                NN = vetorN[id]
                Gera_dados(nome, NN, MM, TT)


def main():
    roda_problemas("CASBTB", 0, 9, "ABB0", 0.9, 1.0, 1.0)
    roda_problemas("CASATB", 0, 9, "AAB0", 0.9, 10.0, 1.0)
    roda_problemas("CASBTA", 0, 9, "ABA0", 0.9, 1.0, 1.5)
    roda_problemas("CASATA", 0, 9, "AAA0", 0.9, 10.0, 1.5)
    roda_problemas("CNSBTB", 0, 9, "NBB0", 1.0, 1.0, 1.0)
    roda_problemas("CNSATB", 0, 9, "NAB0", 1.0, 10.0, 1.0)
    roda_problemas("CNSBTA", 0, 9, "NBA0", 1.0, 1.0, 1.5)
    roda_problemas("CNSATA", 0, 9, "NAA0", 1.0, 10.0, 1.5)


if __name__ == "__main__":
    main()
